### PR TITLE
Add atGivenTimesUnfold Observable #882

### PR DESCRIPTION
**Pull request by RomanIakovlev**
This is an Observable suitable for implementing cron-like schedules,
where tick happens on each predefined instant (with back-pressure applied).

### CODE REVIEW

**Comment by oleg-py** [comment]
Good catch regarding Scala.js. I thought that `java.time` is included in SJS distribution, but apparently it's not.

Let's have a version in shared that takes `Long`s as a timestamp as a base implementation.

Let's also have a platform-specific version for `java.time.Instant` that is only available on JVM and when using Scala 2.12+, on a mixin to `object Observable`.

It has already been done for, e.g., CancelableFuture:
Platform-specific mixins:
https://github.com/monix/monix/blob/master/monix-execution/jvm/src/main/scala_2.11/monix/execution/internal/CancelableFutureForPlatform.scala
https://github.com/monix/monix/blob/master/monix-execution/js/src/main/scala/monix/execution/internal/CancelableFutureForPlatform.scala
https://github.com/monix/monix/blob/master/monix-execution/jvm/src/main/scala_2.12+/monix/execution/internal/CancelableFutureForPlatform.scala

mixin used here:
https://github.com/monix/monix/blob/master/monix-execution/shared/src/main/scala/monix/execution/CancelableFuture.scala#L180

If you feel adventurous, you can add support for java.util.Date (to shared, there's j.u.Date in SJS) and/or sjs native Date.

I also think a version that takes an Iterator or Seq, or something like that makes sense, as that would be more obvious and easy to try out when you aren't using cron4s.

File: monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala [hunk 1]
```diff
@@ -4959,6 +4960,21 @@ object Observable extends ObservableDeprecatedBuilders {
...
+    * `None`, when the resulting Observable completes. Otherwise, its return
+    * value is used to determine next execution instant, as well as input to
+    * its own subsequent invocation.
+    *
+    * @param start input to the `next` function to produce the first tick instant
+    * @param next function emitting the next tick instants based on the result
+    *             of the previous invocation or the `start` parameter
+    * @tparam A type of the elements fed into the `next` function
+    */
+  def atGivenTimesUnfold[A](start: A)(next: A => Option[(A, Instant)]): Observable[Long] =
```
Comment (oleg-py) [inline monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala]: Do you think it makes sense to emit `A` values instead? Could potentially be more general than emitting longs, and longs can be trivially recovered by doing things like `as.zipWithIndex.map(_._2)`.

Comment (RomanIakovlev) [reply monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala]: Agree, changed.

**Comment by RomanIakovlev** [comment]
Ok, I've changed shared code to not depend on `java.time.Instant`. Let's see if tests pass now. I'll add shims for proper types on JS and JVM as a separate commit.

File: monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/AtGivenTimesUnfoldObservable.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+              // relationship between scheduleOnce invocations.
+              prevInstant = nextInstant
+              prevInput = nextInput
+              task := s.scheduleOnce(nextInstant - s.clockRealTime(MILLISECONDS), MILLISECONDS, self)
+            case None =>
+              o.onComplete()
+          }
+        } catch {
+          case NonFatal(ex) =>
+            try {
```
Comment (Avasil) [inline monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/AtGivenTimesUnfoldObservable.scala]: I think this `try catch` is not necessary due to contract:

https://monix.io/docs/3x/reactive/observers.html#contract
(point 7)

File: monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala [hunk 3]
```diff
@@ -4959,6 +4959,20 @@ object Observable extends ObservableDeprecatedBuilders {
...
+    * Observable to complete. Otherwise, its return value is used to
+    * determine next execution instant, as well as input to its own
+    * subsequent invocation.
+    *
+    * @param start input to the `next` function to produce the first tick instant
+    * @param next function emitting the next tick instants based on the result
+    *             of the previous invocation or the `start` parameter
+    */
+  def atGivenTimesUnfold[A](start: A)(next: A => Option[(A, Long)]): Observable[A] =
+    new builders.AtGivenTimesUnfoldObservable[A](start, next)
```
Comment (alexandru) [inline monix-reactive/shared/src/main/scala/monix/reactive/Observable.scala]: This method is pretty cool. Does it do something like this?

```scala
Observable
  .fromAsyncStateAction(Option((start, 0))) { 
    case None => None
    case Some((a, instant)) =>
      Task.deferAction { s =>
        val now = s.clockRealTime(MILLISECONDS)
        val sleep = instant - now
        Task.now(next(a)).delayExecution(sleep.millis)
      }
  }
  .takeWhile(_.nonEmpty)
  .collect { case Some((a, _)) => a }
```

**Comment by alexandru** [comment]
This was a cool suggestion, which we've let open, but I think it can be done with existing operators.

Closing it to clean up the issue tracker.

**Pull request closed without merging**